### PR TITLE
bpo-43987: Add "Annotations Best Practices" HOWTO doc.

### PR DESCRIPTION
**Pull request by larryhastings**
https://bugs.python.org/issue43987

### CODE REVIEW

File: Doc/howto/annotations.rst [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+  Unfortunately, this isn't best practice for classes.  The problem
+  is that, since ``__annotations__`` is optional on classes, and
+  because classes can inherit attributes from their base classes,
+  accessing the ``__annotations__`` attribute of a class may
+  inadvertently return the annotations dict of a *base class.*
+  As an example::
+
+      class Base:
+          a:int = 3
+          b:str = 'abc'
```
Comment (JelleZijlstra) [inline Doc/howto/annotations.rst]:
```suggestion
          a: int = 3
          b: str = 'abc'
```
This is the format recommended by https://www.python.org/dev/peps/pep-0008/#variable-annotations

File: Doc/howto/annotations.rst [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+Manually Un-Stringizing Stringized Annotations
+==============================================
+
+  In situations where some annotations may be "stringized",
+  and you wish to evaluate those strings to produce the
+  Python values they represent, it really is best to
+  call :func:`inspect.get_annotations` to do the evaluations
+  for you.  Un-stringizing is a complicated process, and
+  it's best to let :func:`inspect.get_annotations` do it
+  for you.
```
Comment (JelleZijlstra) [inline Doc/howto/annotations.rst]:
```suggestion
  for you.
```
This sentence is just repetitive

**Review by JelleZijlstra — COMMENTED** [review]
Thanks, this is great! A couple of suggestions:

- It would be useful to add a link in more places to make sure it's findable. You wrote in the issue that "it just didn't seem like a good idea", but I'm not sure why; these other places seem more likely to be seen by users than the glossary or the docs for the new function `inspect.get_annotations`.
- I'm worried that people will come to this page expecting it to be about how to write type hints for their code (there are a lot more people writing type hints than writing code that accesses `__annotations__` at runtime). It'd be good to add links to some document explaining how to do that, perhaps the mypy docs or the docs for the `typing` module.

File: Doc/howto/annotations.rst (hunk 2 above)
Comment (JelleZijlstra) [reply Doc/howto/annotations.rst]: I'm also not sure I agree with the recommendation. There's a few plausible scenarios where it won't work:
- The annotation uses a name defined in a nested scope that's no longer accessible (one of the scenarios that PEP 649 would fix)
- The annotation uses a name defined in `if TYPE_CHECKING`
- The annotation uses PEP 604 syntax `int | str` and `from __future__ import annotations` before Python 3.10

In such cases, eval will throw an error at runtime, but if you try hard enough you can still extract useful information. In my own tool (pyanalyze) I instead `ast.parse` string annotations and evaluate them manually.

I would recommend at least adding a warning that `eval()` may throw for some annotations that static type checkers will process correctly.

Comment (larryhastings) [reply Doc/howto/annotations.rst]: I followed all your suggestions, except adding references from datamodel.rst.  Again, I'm not sure how to work it in in a way that reads well.  Can you give a concrete example of how you'd add a reference to this HOWTO?

p.s. I also found one or two spots where I added a contraction.  When I get exacting, I tend to stop using contractions, and which I think makes my text sound stiff.

p.p.s. Did I forget any quirks?

Comment (JelleZijlstra) [reply Doc/howto/annotations.rst]: > Can you give a concrete example of how you'd add a reference to this HOWTO?

How about adding "For more details on how to use this attribute, see <link>" to the cell for `__annotations__` in the table under https://docs.python.org/3/reference/datamodel.html#the-standard-type-hierarchy ?

**Comment by larryhastings** [comment]
I gave it a shot, see what you think.  I'm still on the fence.  I like having the references there but it still seems to be sticking out a little.  (Then again, `__annotations__` isn't like other dunder attributes, so maybe its asymmetry is inevitable.)

File: Doc/reference/datamodel.rst [hunk 3]
```diff
@@ -750,14 +753,15 @@ Modules
...
-   unavailable; :attr:`__annotations__` (optional) is a dictionary containing
-   :term:`variable annotations <variable annotation>` collected during module
-   body execution; :attr:`__file__` is the pathname of the file from which the
+   unavailable; :attr:`__file__` is the pathname of the file from which the
    module was loaded, if it was loaded from a file. The :attr:`__file__`
    attribute may be missing for certain types of modules, such as C modules
    that are statically linked into the interpreter; for extension modules
    loaded dynamically from a shared library, it is the pathname of the shared
-   library file.
+   library file; :attr:`__annotations__` (optional) is a dictionary containing
```
Comment (JelleZijlstra) [inline Doc/reference/datamodel.rst]: This might work better as a bulleted list, instead of a long string of semicolon-separated fragments.

Comment (larryhastings) [reply Doc/reference/datamodel.rst]: Though I take neither credit nor blame for the original formatting, and such improvements are verging on mission creep, I'll still give it a go.  I'm going to try a "definition list" first and see if I can make that work; failing that, a bulleted list as you suggest.  Either way I'll do that in both spots, modules and classes.

Comment (larryhastings) [reply Doc/reference/datamodel.rst]: I quite like how that turned out!  I invite you to see if you agree.

It makes me wonder if I should now touch up the `Special read-only attribute:` paragraph in the **Modules** section somehow.  On the other hand, there are a bunch of previous instances of *special read-only attributes* sprinkled through the preceding text, so it's not completely out of line.

File: Doc/reference/datamodel.rst [hunk 4]
```diff
@@ -748,16 +751,29 @@ Modules
...
+      :attr:`__doc__`
+         The module's documentation string, or ``None`` if
+         unavailable.
+
+      :attr:`__file__`
+         The pathname of the file from which the
+         module was loaded, if it was loaded from a file.
+         The :attr:`__file__`
+         attribute may be missing for certain types of modules, such as C modules
+         that are statically linked into the interpreter; for extension modules
```
Comment (JelleZijlstra) [inline Doc/reference/datamodel.rst]:
```suggestion
         that are statically linked into the interpreter. For extension modules
```

File: Doc/reference/datamodel.rst [hunk 5]
```diff
@@ -821,14 +837,30 @@ Custom classes
...
-   or ``None`` if undefined; :attr:`__annotations__` (optional) is a dictionary
-   containing :term:`variable annotations <variable annotation>` collected during
-   class body execution.
+   Special attributes:
+
+      :attr:`~definition.__name__`
+         The class name.
+
+      :attr:`__module__`
+         The module name in which the class was defined.
```
Comment (JelleZijlstra) [inline Doc/reference/datamodel.rst]:
```suggestion
         The name of the module in which the class was defined.
```
Classes are defined in modules, not module names.

**Review by JelleZijlstra — COMMENTED** [review]
I agree, this looks a lot more readable! Left a few tiny suggestions.

File: Doc/reference/datamodel.rst (hunk 4 above)
Comment (larryhastings) [reply Doc/reference/datamodel.rst]: Done!

File: Doc/reference/datamodel.rst (hunk 5 above)
Comment (larryhastings) [reply Doc/reference/datamodel.rst]: Done!

**Comment by larryhastings** [comment]
I remembered tonight that "What's New" no longer has a designated curator, but rather is self-serve.  By adding my two recent contributions to (C)Python, I naturally had two more spots I could crowbar in references to "Annotations Best Practices".  See what you think!

File: Doc/whatsnew/3.10.rst [hunk 6]
```diff
@@ -996,7 +998,9 @@ defined on an object. It works around the quirks of accessing the annotations
 on various types of objects, and makes very few assumptions about the object
 it examines. :func:`inspect.get_annotations` can also correctly un-stringize
 stringized annotations.  :func:`inspect.get_annotations` is now considered
-best practice for accessing the annotations dict defined on any Python object.
+best practice for accessing the annotations dict defined on any Python object;
+for more information on best practices for working with annotations, please
```
Comment (JelleZijlstra) [inline Doc/whatsnew/3.10.rst]:
```suggestion
for more information on best practices for working with annotations, please see
```

Comment (larryhastings) [reply Doc/whatsnew/3.10.rst]: Good catch!

**Comment by larryhastings** [comment]
I realize it seems like an unnecessary formality at this point, but I would still prefer (and enjoy!) a LGTM review here.

**Comment by JelleZijlstra** [comment]
I had a comment above on adding more pitfalls about eval that I think would still be useful. I can try to write up some text myself (but not right now).

**Comment by larryhastings** [comment]
Oh, sorry!  Somehow I missed that.  I thought that by adding the word "see" I had addressed all your comments.

I'm about to go to dinner, but when I come back later I'll scroll back and figure it out.  Thanks again for your very thorough reviews!

**Review by JelleZijlstra — APPROVED** [review]
Thanks!

**Comment by larryhastings** [comment]
Thanks for all your time on this, Jelle!

**Pull request merged**